src/evaluators/CallExpression.js

```diff
@@ -98,43 +98,85 @@ function callBothFunctionsAndJoinTheirEffects(
   return completion;
 }
 
-function EvaluateCall(
+function generateRuntimeCall(
   ref: Value | Reference,
   func: Value,
   ast: BabelNodeCallExpression,
   strictCode: boolean,
   env: LexicalEnvironment,
   realm: Realm
-): Value {
-  function generateRuntimeCall() {
-    let args = [func];
-    let [thisArg, propName] = ref instanceof Reference ? [ref.base, ref.referencedName] : [];
-    if (thisArg instanceof Value) args = [thisArg];
-    if (propName !== undefined && typeof propName !== "string") args.push(propName);
-    args = args.concat(ArgumentListEvaluation(realm, strictCode, env, ast.arguments));
-    for (let arg of args) {
-      if (arg !== func) {
-        Leak.leakValue(realm, arg, ast.loc);
-      }
+) {
+  let args = [func];
+  let [thisArg, propName] = ref instanceof Reference ? [ref.base, ref.referencedName] : [];
+  if (thisArg instanceof Value) args = [thisArg];
+  if (propName !== undefined && typeof propName !== "string") args.push(propName);
+  args = args.concat(ArgumentListEvaluation(realm, strictCode, env, ast.arguments));
+  for (let arg of args) {
+    if (arg !== func) {
+      Leak.leakValue(realm, arg, ast.loc);
     }
-    return AbstractValue.createTemporalFromBuildFunction(realm, Value, args, nodes => {
-      let callFunc;
-      let argStart = 1;
-      if (thisArg instanceof Value) {
-        if (typeof propName === "string") {
-          callFunc = t.memberExpression(nodes[0], t.identifier(propName), !t.isValidIdentifier(propName));
-        } else {
-          callFunc = t.memberExpression(nodes[0], nodes[1], true);
-          argStart = 2;
-        }
+  }
+  return AbstractValue.createTemporalFromBuildFunction(realm, Value, args, nodes => {
+    let callFunc;
+    let argStart = 1;
+    if (thisArg instanceof Value) {
+      if (typeof propName === "string") {
+        callFunc = t.memberExpression(nodes[0], t.identifier(propName), !t.isValidIdentifier(propName));
       } else {
-        callFunc = nodes[0];
+        callFunc = t.memberExpression(nodes[0], nodes[1], true);
+        argStart = 2;
       }
-      let fun_args = ((nodes.slice(argStart): any): Array<BabelNodeExpression | BabelNodeSpreadElement>);
-      return t.callExpression(callFunc, fun_args);
-    });
+    } else {
+      callFunc = nodes[0];
+    }
+    let fun_args = ((nodes.slice(argStart): any): Array<BabelNodeExpression | BabelNodeSpreadElement>);
+    return t.callExpression(callFunc, fun_args);
+  });
+}
+
+function evaluatePureCall(
+  ref: Value | Reference,
+  func: Value,
+  ast: BabelNodeCallExpression,
+  strictCode: boolean,
+  env: LexicalEnvironment,
+  realm: Realm,
+  thisValue: Value,
+  tailCall: boolean
+): Value {
+  let effects;
+  try {
+    effects = realm.evaluateForEffects(() =>
+      EvaluateDirectCall(realm, strictCode, env, ref, func, thisValue, ast.arguments, tailCall)
+    );
+  } catch (error) {
+    if (error instanceof FatalError) {
```

Comment:
@hermanventer Does this check seem ok to you?

The idea here is that we can optimistically evaluate something but that causes a prepack error (as opposed to a throw completion), we can fallback to treating the whole call as abstract.